core/src/com/unciv/logic/map/tile/Tile.kt

```diff
@@ -852,15 +872,39 @@ class Tile : IsPartOfGameInfoSerialization {
     }
 
     fun startWorkingOnImprovement(improvement: TileImprovement, civInfo: Civilization, unit: MapUnit) {
-        improvementInProgress = improvement.name
-        turnsToImprovement = if (civInfo.gameInfo.gameParameters.godMode) 1
-            else improvement.getTurnsToBuild(civInfo, unit)
+        improvementQueue.clear()
+        queueImprovement(improvement, civInfo, unit)
     }
 
-    /** Clears [improvementInProgress] and [turnsToImprovement] */
+    /** Clears [improvementQueue] */
     fun stopWorkingOnImprovement() {
-        improvementInProgress = null
-        turnsToImprovement = 0
+        improvementQueue.clear()
+    }
+
+    /** Adds an entry to the [improvementQueue], by looking up the time it takes using [civInfo] and [unit] */
+    fun queueImprovement(improvement: TileImprovement, civInfo: Civilization, unit: MapUnit) {
+        val turns = if (civInfo.gameInfo.gameParameters.godMode) 1
+            else improvement.getTurnsToBuild(civInfo, unit)
+        queueImprovement(improvement.name, turns)
+    }
+
+    /** Adds an entry to the [improvementQueue] with explicit [turnsToImprovement] */
+    fun queueImprovement(improvementName: String, turnsToImprovement: Int) {
+        improvementQueue.add(ImprovementQueueEntry(improvementName, turnsToImprovement))
+    }
+
+    /** Called from UnitTurnManager when a Worker "spends time" here */
+    fun endTurn(worker: MapUnit) {
```

Comment:
This should be renamed to something more like `doWorkerTurn` to clarify that we are doing improvement business.